src/hazelcore/HTyp.re

```diff
@@ -42,13 +49,24 @@ let precedence = (ty: t): int =>
 /* equality
    At the moment, this coincides with default equality,
    but this will change when polymorphic types are implemented */
-let eq = (==);
+let eq = (x, y) =>
+  switch (x, y) {
+  | (TyVarHole(_), TyVarHole(_))
```

Comment:
we shouldn't call all TyVarHoles equal to each other or to Holes. they should only be equal if they are the same variable.